cpp/include/cudf/ast/detail/expression_parser.hpp

```diff
@@ -248,12 +250,17 @@ class expression_parser {
     extract_size_and_pointer(_data_references, sizes, data_pointers);
     extract_size_and_pointer(_literals, sizes, data_pointers);
     extract_size_and_pointer(_operators, sizes, data_pointers);
+    extract_size_and_pointer(_operator_arities, sizes, data_pointers);
     extract_size_and_pointer(_operator_source_indices, sizes, data_pointers);
 
     // Create device buffer
     auto const buffer_size = std::accumulate(sizes.cbegin(), sizes.cend(), 0);
     auto buffer_offsets    = std::vector<int>(sizes.size());
-    thrust::exclusive_scan(sizes.cbegin(), sizes.cend(), buffer_offsets.begin(), 0);
+    thrust::exclusive_scan(
+      sizes.cbegin(), sizes.cend(), buffer_offsets.begin(), 0, [](auto a, auto b) {
+        // Must align each part of the AST program on 4-byte addresses
```

Comment:
@lamarrr I tried to make the arity an int8_t instead of an int32_t but ran into memory access issues, which I guessed were from address alignment. I tried to work around it by adding this but it didn’t seem to work, so I am not sure if it’s still needed. We should try reverting this and/or figure out why packing an int8_t wasn’t working. I think we would have to read on 4-byte aligned addresses and doing that with int8_t would required masking out the desired bits from that. I don’t love that because it’s extra code, so maybe just using int32_t and removing this manual alignment is the right approach.

Reply:
yeah, using an int32_t would be the right approach, SMPs don't work well with lesser than 16-byte reads let alone 1-byte.

Reply:
I've created an MR to your branch that attempts to fix this. It should be fixed by that. The total buffer size was being computed using the previous unaligned size